scoped-mutex/src/lib.rs

```diff
@@ -0,0 +1,125 @@
+#![no_std]
+
+use core::cell::UnsafeCell;
+use core::sync::atomic::{AtomicBool, Ordering};
+
+/// Raw mutex trait.
+///
+/// This mutex is "raw", which means it does not actually contain the protected data, it
+/// just implements the mutex mechanism. For most uses you should use [`super::Mutex`] instead,
+/// which is generic over a RawMutex and contains the protected data.
+///
+/// Note that, unlike other mutexes, implementations only guarantee no
+/// concurrent access from other threads: concurrent access from the current
+/// thread is allowed. For example, it's possible to lock the same mutex multiple times reentrantly.
+///
+/// Therefore, locking a `RawMutex` is only enough to guarantee safe shared (`&`) access
+/// to the data, it is not enough to guarantee exclusive (`&mut`) access.
+///
+/// # Safety
+///
+/// RawMutex implementations must ensure that, while locked, no other thread can lock
+/// the RawMutex concurrently.
+///
+/// Unsafe code is allowed to rely on this fact, so incorrect implementations will cause undefined behavior.
+pub unsafe trait RawMutex {
+    /// Create a new `RawMutex` instance.
+    ///
+    /// This is a const instead of a method to allow creating instances in const context.
+    const INIT: Self;
```

Comment:
This is a hard blocker for `maitake-sync` , unfortunately. In `maitake-sync`, I use `loom` for testing, and `loom`'s simulated mutex cannot be instantiated in a `const` context, because it must be bound to the _current iteration_ of the Loom test at runtime. I need to be able to implement the trait that has the core `Mutex` functionality for the `loom::sync::Mutex` type, so it can't require a const initializer.

Instead, I think the best approach would be to factor this out into its own trait, like this:
```rust
pub unsafe trait RawMutex {
    /// Lock this `RawMutex`.
    fn lock<R>(&self, f: impl FnOnce() -> R) -> R;
}

pub trait ConstInit {
    const INIT: Self;
}
```

This way, `Mutex` wrapper types like `BlockingMutex` can have `const fn` constructors`where R: RawMutex + ConstInit`, but there can still be a way to construct them with non-const raw mutices (possibly `Default`?).

(as a side note, my preference would be to _not_ make the `ConstInit` trait have a `Self: RawMutex` bound, as I'd like to be able to reuse it for my `RawRwLock` trait, but if it needs to, I can live with that)